programs/leap-util/actions/snapshot.cpp

```diff
@@ -10,29 +16,86 @@
 #include <boost/filesystem.hpp>
 #include <boost/filesystem/path.hpp>
 
+using namespace eosio;
+using namespace eosio::chain;
+
+namespace bfs = boost::filesystem;
+
 void snapshot_actions::setup(CLI::App& app) {
-   auto* sub = app.add_subcommand("snapshot", "snapshot utility");
-   sub->add_subcommand("to-json", "convert snapshot file to convert to json format");
+   auto* sub = app.add_subcommand("snapshot", "Snapshot utility");
+   sub->require_subcommand(1);
 
-   // options
-   sub->add_option("--input-file,-i", opt->input_file, "snapshot file to convert to json format, writes to <file>.json if output file not specified (tmp state dir used), and exit.")->required();
-   sub->add_option("--output-file,-o", opt->output_file, "the file to write the output to (absolute or relative path).  if not specified then output is to stdout.");
+   // subcommand -convert snapshot to json
+   auto to_json = sub->add_subcommand("to-json", "Convert snapshot file to json format");
+   to_json->add_option("--input-file,-i", opt->input_file, "Snapshot file to convert to json format, writes to <file>.json if output file not specified (tmp state dir used).")->required();
+   to_json->add_option("--output-file,-o", opt->output_file, "The file to write the output to (absolute or relative path).  If not specified then output is to <input-file>.json.");
 
-   sub->callback([this]() {
-      int rc = run_subcommand();
-      // properly return err code in main
-      if(rc) throw(CLI::RuntimeError(rc));
+   to_json->callback([this]() {
+      try {
+         int rc = run_subcommand();
+         if(rc) throw(CLI::RuntimeError(rc));
+      } catch(...) {
+         print_exception();
+         throw(CLI::RuntimeError(-1));
+      }
    });
 }
 
 int snapshot_actions::run_subcommand() {
    if(!opt->input_file.empty()) {
       if(!fc::exists(opt->input_file)) {
-         std::cerr << "cannot load snapshot, " << opt->input_file << " does not exist" << std::endl;
+         std::cerr << "cannot load snapshot, " << opt->input_file
+                   << " does not exist" << std::endl;
          return -1;
       }
    }
-   // todo: implementation here
 
+   bfs::path snapshot_path = opt->input_file;
+   bfs::path json_path = opt->output_file.empty()
+                               ? snapshot_path.generic_string() + ".json"
+                               : opt->output_file;
+   // pull chain id
+   auto infile = std::ifstream(snapshot_path.generic_string(),
+                               (std::ios::in | std::ios::binary));
+   istream_snapshot_reader reader(infile);
+   reader.validate();
+   auto chain_id = controller::extract_chain_id(reader);
+   infile.close();
```

Comment:
What if chain id is not in the snapshot? snapshot version < 2.0. Should we have an option to specify `chain_id`?

Reply:
done